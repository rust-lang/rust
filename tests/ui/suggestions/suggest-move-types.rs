#![allow(warnings)]

// This test verifies that the suggestion to move types before associated type bindings
// is correct.

trait One<T> {
  type A;
}

trait OneWithLifetime<'a, T> {
  type A;
}

trait Three<T, U, V> {
  type A;
  type B;
  type C;
}

trait ThreeWithLifetime<'a, 'b, 'c, T, U, V> {
  type A;
  type B;
  type C;
}

struct A<T, M: One<A=(), T>> {
//~^ ERROR generic arguments must come before the first constraint
    m: M,
    t: T,
}


struct Al<'a, T, M: OneWithLifetime<A=(), T, 'a>> {
//~^ ERROR generic arguments must come before the first constraint
//~^^ ERROR type provided when a lifetime was expected
    m: M,
    t: &'a T,
}

struct B<T, U, V, M: Three<A=(), B=(), C=(), T, U, V>> {
//~^ ERROR generic arguments must come before the first constraint
    m: M,
    t: T,
    u: U,
    v: V,
}

struct Bl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<A=(), B=(), C=(), T, U, V, 'a, 'b, 'c>> {
//~^ ERROR generic arguments must come before the first constraint
//~^^ ERROR type provided when a lifetime was expected
    m: M,
    t: &'a T,
    u: &'b U,
    v: &'c V,
}

struct C<T, U, V, M: Three<T, A=(), B=(), C=(), U, V>> {
//~^ ERROR generic arguments must come before the first constraint
    m: M,
    t: T,
    u: U,
    v: V,
}

struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), C=(), U, 'b, V, 'c>> {
//~^ ERROR generic arguments must come before the first constraint
//~^^ ERROR lifetime provided when a type was expected
    m: M,
    t: &'a T,
    u: &'b U,
    v: &'c V,
}

struct D<T, U, V, M: Three<T, A=(), B=(), U, C=(), V>> {
//~^ ERROR generic arguments must come before the first constraint
    m: M,
    t: T,
    u: U,
    v: V,
}

struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, 'b, C=(), V, 'c>> {
//~^ ERROR generic arguments must come before the first constraint
//~^^ ERROR lifetime provided when a type was expected
    m: M,
    t: &'a T,
    u: &'b U,
    v: &'c V,
}

fn main() {}
