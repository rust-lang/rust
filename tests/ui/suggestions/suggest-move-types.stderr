error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:26:26
   |
LL | struct A<T, M: One<A=(), T>> {
   |                    ----  ^ generic argument
   |                    |
   |                    constraint
   |
help: move the constraint after the generic argument
   |
LL | struct A<T, M: One<T, A = ()>> {
   |                   ~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:33:43
   |
LL | struct Al<'a, T, M: OneWithLifetime<A=(), T, 'a>> {
   |                                     ----  ^  ^^ generic arguments
   |                                     |
   |                                     constraint
   |
help: move the constraint after the generic arguments
   |
LL | struct Al<'a, T, M: OneWithLifetime<'a, T, A = ()>> {
   |                                    ~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:40:46
   |
LL | struct B<T, U, V, M: Three<A=(), B=(), C=(), T, U, V>> {
   |                            ----  ----  ----  ^  ^  ^ generic arguments
   |                            |
   |                            constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct B<T, U, V, M: Three<T, U, V, A = (), B = (), C = ()>> {
   |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:48:71
   |
LL | struct Bl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<A=(), B=(), C=(), T, U, V, 'a, 'b, 'c>> {
   |                                                     ----  ----  ----  ^  ^  ^  ^^  ^^  ^^ generic arguments
   |                                                     |
   |                                                     constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct Bl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<'a, 'b, 'c, T, U, V, A = (), B = (), C = ()>> {
   |                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:57:28
   |
LL | struct C<T, U, V, M: Three<T, A=(), B=(), C=(), U, V>> {
   |                            ^  ----  ----  ----  ^  ^ generic arguments
   |                               |
   |                               constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct C<T, U, V, M: Three<T, U, V, A = (), B = (), C = ()>> {
   |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:65:53
   |
LL | struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), C=(), U, 'b, V, 'c>> {
   |                                                     ^  ^^  ----  ----  ----  ^  ^^  ^  ^^ generic arguments
   |                                                            |
   |                                                            constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<'a, 'b, 'c, T, U, V, A = (), B = (), C = ()>> {
   |                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:74:28
   |
LL | struct D<T, U, V, M: Three<T, A=(), B=(), U, C=(), V>> {
   |                            ^  ----  ----  ^  ----  ^ generic arguments
   |                               |
   |                               constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct D<T, U, V, M: Three<T, U, V, A = (), B = (), C = ()>> {
   |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: generic arguments must come before the first constraint
  --> $DIR/suggest-move-types.rs:82:53
   |
LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, 'b, C=(), V, 'c>> {
   |                                                     ^  ^^  ----  ----  ^  ^^  ----  ^  ^^ generic arguments
   |                                                            |
   |                                                            constraints
   |
help: move the constraints after the generic arguments
   |
LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<'a, 'b, 'c, T, U, V, A = (), B = (), C = ()>> {
   |                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0747]: type provided when a lifetime was expected
  --> $DIR/suggest-move-types.rs:33:43
   |
LL | struct Al<'a, T, M: OneWithLifetime<A=(), T, 'a>> {
   |                                           ^

error[E0747]: type provided when a lifetime was expected
  --> $DIR/suggest-move-types.rs:48:71
   |
LL | struct Bl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<A=(), B=(), C=(), T, U, V, 'a, 'b, 'c>> {
   |                                                                       ^

error[E0747]: lifetime provided when a type was expected
  --> $DIR/suggest-move-types.rs:65:56
   |
LL | struct Cl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), C=(), U, 'b, V, 'c>> {
   |                                                        ^^
   |
   = note: lifetime arguments must be provided before type arguments
   = help: reorder the arguments: lifetimes, then type and consts: `<'a, 'b, 'c, T, U, V>`

error[E0747]: lifetime provided when a type was expected
  --> $DIR/suggest-move-types.rs:82:56
   |
LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, 'b, C=(), V, 'c>> {
   |                                                        ^^
   |
   = note: lifetime arguments must be provided before type arguments
   = help: reorder the arguments: lifetimes, then type and consts: `<'a, 'b, 'c, T, U, V>`

error: aborting due to 12 previous errors

For more information about this error, try `rustc --explain E0747`.
